File: triforce/server/protocol.py
"""DANTE UI wire protocol v1 — Python mirror of ui/web/src/protocol.ts.

Pure stdlib: no fastapi/adk imports here, so this module always imports.
The contract lives in ui/ARCHITECTURE.md §2; change that first.

Receiver rules (both directions):
  - Unknown ``type``            -> log + drop (no reply).
  - Unknown fields on known types -> ignored (forward compatibility).
  - ``v`` > PROTOCOL_VERSION     -> log + drop.
  - Malformed JSON / bad shape   -> ``error`` with code ``bad_message``.
"""

from __future__ import annotations

import json
from typing import Any, Optional

PROTOCOL_VERSION = 1

AGENT_STATES = ("idle", "listening", "thinking", "speaking")
SYSTEM_MODES = ("awake", "sleep", "reflective")
ERROR_CODES = ("bad_message", "executor_unavailable", "internal")

CLIENT_TYPES = ("client.hello", "chat.send", "chat.cancel", "ping")

# ---------------------------------------------------------------------------
# Server -> Client message builders
# ---------------------------------------------------------------------------


def server_hello(simulated: bool) -> dict:
    return {
        "v": PROTOCOL_VERSION,
        "type": "server.hello",
        "protocol": PROTOCOL_VERSION,
        "simulated": bool(simulated),
    }


def chat_delta(exchange_id: str, text: str) -> dict:
    return {"v": PROTOCOL_VERSION, "type": "chat.delta", "id": exchange_id, "text": text}


def chat_complete(exchange_id: str) -> dict:
    return {"v": PROTOCOL_VERSION, "type": "chat.complete", "id": exchange_id}


def agent_state(state: str) -> dict:
    assert state in AGENT_STATES, state
    return {"v": PROTOCOL_VERSION, "type": "agent.state", "state": state}


def system_mode(mode: str) -> dict:
    assert mode in SYSTEM_MODES, mode
    return {"v": PROTOCOL_VERSION, "type": "system.mode", "mode": mode}


def error(code: str, message: str, exchange_id: Optional[str] = None) -> dict:
    assert code in ERROR_CODES, code
    msg: dict[str, Any] = {
        "v": PROTOCOL_VERSION,
        "type": "error",
        "code": code,
        "message": message,
    }
    if exchange_id is not None:
        msg["id"] = exchange_id
    return msg


def pong() -> dict:
    return {"v": PROTOCOL_VERSION, "type": "pong"}


# ---------------------------------------------------------------------------
# Client -> Server parsing / validation
# ---------------------------------------------------------------------------


class ParseResult:
    """Outcome of parsing one inbound frame.

    Exactly one of the three shapes:
      - ``message`` set        -> valid, dispatch it.
      - ``reply_error`` set    -> send that ``error`` message back.
      - ``drop_reason`` set    -> log and silently drop (unknown type / higher v).
    """

    __slots__ = ("message", "reply_error", "drop_reason")

    def __init__(
        self,
        message: Optional[dict] = None,
        reply_error: Optional[dict] = None,
        drop_reason: Optional[str] = None,
    ):
        self.message = message
        self.reply_error = reply_error
        self.drop_reason = drop_reason


_REQUIRED_STR_FIELDS = {
    "client.hello": ("client",),
    "chat.send": ("id", "text"),
    "chat.cancel": ("id",),
    "ping": (),
}


def parse_client_frame(raw: str) -> ParseResult:
    """Validate one raw text frame against protocol v1."""
    try:
        msg = json.loads(raw)
    except (ValueError, TypeError):
        return ParseResult(reply_error=error("bad_message", "frame is not valid JSON"))

    if not isinstance(msg, dict):
        return ParseResult(reply_error=error("bad_message", "frame is not a JSON object"))

    v = msg.get("v")
    mtype = msg.get("type")

    if not isinstance(v, int) or not isinstance(mtype, str):
        return ParseResult(
            reply_error=error("bad_message", "missing or invalid envelope fields 'v'/'type'")
        )

    if v > PROTOCOL_VERSION:
        return ParseResult(drop_reason=f"message v={v} > supported {PROTOCOL_VERSION}")

    if mtype not in _REQUIRED_STR_FIELDS:
        return ParseResult(drop_reason=f"unknown message type {mtype!r}")

    for field in _REQUIRED_STR_FIELDS[mtype]:
        value = msg.get(field)
        if not isinstance(value, str) or not value:
            exchange_id = msg.get("id") if isinstance(msg.get("id"), str) else None
            return ParseResult(
                reply_error=error(
                    "bad_message",
                    f"{mtype} requires string field {field!r}",
                    exchange_id,
                )
            )

    return ParseResult(message=msg)

File: ui/web/src/protocol.ts
/**
 * DANTE UI wire protocol, v1.
 *
 * EXACT mirror of ui/ARCHITECTURE.md §2 "Protocol spec (v1)" — that document
 * is the contract; change it first, then this file, in the same commit.
 *
 * Transport: single WebSocket, UTF-8 JSON text frames, one message per frame.
 * Endpoint:  ws://127.0.0.1:8420/ws
 *
 * Rules for receivers (both directions):
 *  - Ignore (log + drop) messages with unknown `type`.
 *  - Ignore unknown fields on known types (forward compatibility).
 *  - Ignore messages whose `v` is greater than PROTOCOL_VERSION (log + drop).
 */

export const PROTOCOL_VERSION = 1 as const;

export const WS_URL = 'ws://127.0.0.1:8420/ws' as const;
export const HEALTH_URL = 'http://127.0.0.1:8420/healthz' as const;

/** Avatar/brain animation state, pushed by the server on every change. */
export type AgentState = 'idle' | 'listening' | 'thinking' | 'speaking';

/** DANTE system mode, pushed by the server on every change. */
export type SystemMode = 'awake' | 'sleep' | 'reflective';

/** Kind of client connecting, announced in `client.hello`. */
export type ClientKind = 'web' | 'desktop';

/** Recoverable error codes carried by `error` messages. */
export type ErrorCode = 'bad_message' | 'executor_unavailable' | 'internal';

/** Common envelope: every message has `v` and a `type` discriminator. */
export interface Envelope {
  v: typeof PROTOCOL_VERSION;
  type: string;
}

/* ------------------------------------------------------------------ */
/* Client → Server                                                     */
/* ------------------------------------------------------------------ */

/** First message after connect. Announces client kind and protocol version. */
export interface ClientHello extends Envelope {
  type: 'client.hello';
  client: ClientKind;
}

/** User sends a chat message. `id` is a client-minted UUID v4. */
export interface ChatSend extends Envelope {
  type: 'chat.send';
  id: string;
  text: string;
}

/**
 * Ask the server to stop generating the reply for `id`. Server responds with
 * `chat.complete` for that `id` (possibly after a final partial delta).
 */
export interface ChatCancel extends Envelope {
  type: 'chat.cancel';
  id: string;
}

/** Liveness probe; server answers `pong`. Sent every 20 s of socket idle. */
export interface Ping extends Envelope {
  type: 'ping';
}

export type ClientMessage = ClientHello | ChatSend | ChatCancel | Ping;

/* ------------------------------------------------------------------ */
/* Server → Client                                                     */
/* ------------------------------------------------------------------ */

/**
 * First message after connect. `simulated: true` means the real Executor
 * (google-adk/LLM) is unavailable and replies come from the simulated
 * Executor — the UI should show a subtle "demo mode" indicator.
 */
export interface ServerHello extends Envelope {
  type: 'server.hello';
  protocol: number;
  simulated: boolean;
}

/** One streamed token/chunk of the assistant reply for exchange `id`. */
export interface ChatDelta extends Envelope {
  type: 'chat.delta';
  id: string;
  text: string;
}

/**
 * Reply for `id` is finished (normally or via cancel). No further
 * `chat.delta` for this `id` will follow.
 */
export interface ChatComplete extends Envelope {
  type: 'chat.complete';
  id: string;
}

/**
 * Avatar/brain animation state. Pushed on every change, and pushed once
 * (current value) right after `server.hello`.
 */
export interface AgentStateMsg extends Envelope {
  type: 'agent.state';
  state: AgentState;
}

/** DANTE system mode. Pushed on change and right after `server.hello`. */
export interface SystemModeMsg extends Envelope {
  type: 'system.mode';
  mode: SystemMode;
}

/**
 * Future: Temporal-driven dream/reflection notification. Clients MAY ignore
 * in v1; defined now so the envelope is stable.
 */
export interface DreamNotify extends Envelope {
  type: 'dream.notify';
  title: string;
  body: string;
}

/**
 * Recoverable error. `id` present when tied to a chat exchange (that
 * exchange is then terminated — treat as `chat.complete`).
 */
export interface ErrorMsg extends Envelope {
  type: 'error';
  code: ErrorCode;
  message: string;
  id?: string;
}

/** Reply to `ping`. */
export interface Pong extends Envelope {
  type: 'pong';
}

export type ServerMessage =
  | ServerHello
  | ChatDelta
  | ChatComplete
  | AgentStateMsg
  | SystemModeMsg
  | DreamNotify
  | ErrorMsg
  | Pong;

/* ------------------------------------------------------------------ */
/* Reconnect policy (client-side, see ARCHITECTURE.md)                 */
/* ------------------------------------------------------------------ */

/** Exponential backoff schedule: 500 ms doubling to an 8 s cap, ±20% jitter. */
export const RECONNECT_BASE_MS = 500;
export const RECONNECT_MAX_MS = 8_000;
export const RECONNECT_JITTER = 0.2;
/** A connection surviving this long resets the backoff to base. */
export const RECONNECT_STABLE_MS = 10_000;
/** Client sends `ping` after this much socket idle. */
export const PING_IDLE_MS = 20_000;
/**
 * After sending `ping`, some inbound frame (normally `pong`) must arrive
 * within this window or the socket is treated as half-open/dead and is
 * force-closed so the normal reconnect + demo-mode fallback take over.
 */
export const PONG_TIMEOUT_MS = 10_000;
